compiler/rustc_error_codes/src/error_codes/E0038.md

```diff
@@ -1,34 +1,65 @@
-Trait objects like `Box<Trait>` can only be constructed when certain
-requirements are satisfied by the trait in question.
+For any given trait `Trait` there may be a related _type_ called the _trait
+object type_ which is typically written as `dyn Trait`. Some traits are not
+allowed to be used as trait object types.
 
-Trait objects are a form of dynamic dispatch and use a dynamically sized type
-for the inner type. So, for a given trait `Trait`, when `Trait` is treated as a
-type, as in `Box<Trait>`, the inner type is 'unsized'. In such cases the boxed
-pointer is a 'fat pointer' that contains an extra pointer to a table of methods
-(among other things) for dynamic dispatch. This design mandates some
-restrictions on the types of traits that are allowed to be used in trait
-objects, which are collectively termed as 'object safety' rules.
+The traits that are allowed to be used as trait object types are called
+"object-safe" traits.  Attempting to use a trait object type for a trait that is
+not object-safe will trigger error E0038.
 
-Attempting to create a trait object for a non object-safe trait will trigger
-this error.
+In earlier editions of Rust, trait object types were written as plain `Trait`
```

Comment:
I wonder if this sentence should move into the introduction or right after where `dyn Trait` is mentioned for the first time? As it is, it feels a bit like it's spliced between parts of the "what is/is not object safe" discussion.

Reply:
Something like:

> For any given trait `Trait` there may be a related _type_ called the _trait
object type_ which is typically written as `dyn Trait`. In earlier editions of Rust, 
trait object types were written as plain `Trait` (just the name of the trait, written in 
type positions) but this was a bit too confusing, so we now write `dyn Trait`.
> 
> Not all traits are allowed to be used as trait object types. The traits that are allowed 
to be used as trait object types are called "object-safe" traits. Attempting to use a trait
object type for a trait that is not object-safe will trigger error E0038.
>
> Two general aspects of trait object types give rise to the restrictions:
> ...

Flows a bit better IMO but this is just a suggestion 🙂

Reply:
I had it like that at first, but then I rewrote to the current form because I wanted to preserve the one nice thing about the existing docs, which is that it says "why you got this error" in a hurry -- "not all traits are allowed to be used as object types". That's the meat of the explanation.

But sure, if you like I'm happy to switch it back. I agree that it's a little more connected to the content in the first paragraph, just makes it longer and delays getting to the point.

Reply:
Yeah, that's a good point. A quick look at some of the other errors in this file suggests we don't follow a consistent pattern; some errors get straight to the point while others have more upfront exposition. 

I think either version is better than what we currently have so I'm going to r+. If you feel strongly that you like your original version better, I'm also ok with landing that instead of the current version. Just let me know if you want to do that.